### PR TITLE
feat: Store keys upon put

### PR DESCRIPTION
**Pull request by stanleyjones**
## Description

This PR should unblock liftedinit/roadmap#17 by populating a list of keys that can be fetched as the user writes them.

## Related Issue

Fixes https://github.com/liftedinit/roadmap/issues/17 (But doesn't really)

## Testing

All tests pass.

## Breaking Changes (if applicable)

None.

## Screenshots (if applicable)

Screenshot 2023-01-23 at 10 37 40 AM

## Checklist:

- [x] I have read and followed the CONTRIBUTING guidelines for this project.
- [x] I have added or updated tests and they pass.
- [x] I have added or updated documentation and it is accurate.
- [x] I have noted any breaking changes in this module or downstream modules.

### CODE REVIEW

File: src/features/services/components/data-settings.tsx [hunk 1]
```diff
@@ -34,6 +35,8 @@ function KVDataRow({ key, value, tag }: KVData) {
 
 export function DataSettings() {
   const account = useAccountsStore((s) => s.byId.get(s.activeId));
+  // @TODO: Fetch all the keys
+  const keys = useDataServiceStore((s) => s.keys);
```
Comment (stanleyjones) [inline src/features/services/components/data-settings.tsx:39]: Here's how you can get the keys out of state.

File: src/features/services/components/data-settings.tsx [hunk 2]
```diff
@@ -54,12 +57,12 @@ export function DataSettings() {
...
-            Create New Key
-          </Button>
-        </Flex>
-      )}
+          <Flex mt={9} justifyContent="flex-end" w="full">
+            <Button width={{ base: "full", md: "auto" }} onClick={onOpen}>
+              Create New Key
+            </Button>
+          </Flex>
+        )}
```
Comment (stanleyjones) [inline src/features/services/components/data-settings.tsx:65]: Just indenting.

File: src/features/services/queries.ts [hunk 3]
```diff
@@ -94,15 +98,19 @@ export function usePutValue() {
...
   const queryClient = useQueryClient();
+  const addKey = useDataServiceStore((s) => s.addKey);
 
   return useMutation<unknown, Error, PutValueInputs>({
     mutationFn: async (inputs: PutValueInputs) => {
       await network?.kvStore.put(inputs);
     },
-    onSuccess: () =>
+    onSuccess: (_, inputs) => {
+      addKey(inputs.key);
```
Comment (stanleyjones) [inline src/features/services/queries.ts:108]: Add the key to local state if we successfully wrote it to the back-end.

File: src/features/services/store.ts [hunk 4]
```diff
@@ -111,3 +114,32 @@ export const useServicesStore = create<ServicesState & ServicesActions>(
...
+>(
+  persist(
+    (set) => ({
+      keys: [],
+      addKey: (key: string) => set((s) => ({ keys: [...s.keys, key] })),
+      delKey: (key: string) =>
+        set((s) => ({ keys: s.keys.filter((k) => k !== key) })),
+    }),
+    {
+      name: "GWEN.SERVICES.DATA",
```
Comment (stanleyjones) [inline src/features/services/store.ts:138]: You can look in the browser storage under this key to confirm keys are being written.

**Comment by stanleyjones** [comment]
Hmm... CI failed. Investigating.

**Comment by stanleyjones** [comment]
Working now.

**Review by mjsorribas — APPROVED** [review]

**Comment by stanleyjones** [comment]
Oh, this is already merged. Okay.

I think you'll have to copy the changes from https://github.com/stanleyjones/gwen/pull/1 into your own branch then.

**Pull request merged**